File: Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.py
class Solution:
    def addToDict(self, ht, ch):
        if (ch not in ht):
            ht[ch] = 1
        else:
            ht[ch] += 1
        
    def removeFromDict(self, ht, ch):
        ht[ch] -= 1
    
    def isValid(self, ht):
        for ch in ht:
            if (ht[ch] > 1):
                return False
        return True
    
    
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        fp = 0
        sp = 0
        ht = { }
        ans = 0
        while (sp < n):
            self.addToDict(ht, s[sp])
            while (fp < sp and not self.isValid(ht)):
                self.removeFromDict(ht, s[fp])
                fp += 1
            
            length = sp - fp + 1
            ans = max(ans, length)
            sp += 1
        
        return ans
            
    

File: Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> mp;
        int ans = 0;
        int fp = 0, sp = 0, n = s.length();
        
        while (sp < n) {
            mp[s[sp]]++;
            
            while (fp < sp && !isValid(mp)) {
                mp[s[fp]]--;
                fp++;
            }
            
            int length = sp - fp + 1;
            if (length > ans) {
                ans = length;
            }
            sp++;
        }
        
        return ans;
    }
    
    
    bool isValid(unordered_map<char, int>& mp) {
        for(auto it = mp.begin(); it != mp.end(); it++) {
            if (it->second > 1) {
                return false;
            }
        }
        return true;
    }
};